packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/queue/queue.ts

```diff
@@ -0,0 +1,389 @@
+}
+
+/**
+ * Options for configuring SQS Queue auto scaling.
+ */
+interface QueueAutoScalingOptions {
+  /**
+   * Acceptable amount of time taken to process one message in the queue.
+   */
+  readonly messageProcessingTime: cdk.Duration;
+
+  /**
+   * Acceptable amount of time for processing a single message in the queue.
+   */
+  readonly acceptableLatency: cdk.Duration;
+}
+
+/**
+ * The `TopicSubscription` class represents an SNS Topic resource that can be subscribed to by the service queues.
+ */
+export class TopicSubscription implements ISubscribable {
+  public readonly topic: sns.ITopic;
+
+  public readonly queue?: sqs.IQueue;
+
+  private readonly subscriptionQueue?: SubscriptionQueue;
+
+  constructor(props: TopicSubscriptionProps) {
+    this.topic = props.topic;
+    this.subscriptionQueue = props.topicSubscriptionQueue;
+    this.queue = props.topicSubscriptionQueue?.queue;
+  }
+
+  /**
+   * This method sets up SNS Topic subscriptions for the SQS queue provided by the user. If a `queue` is not provided,
+   * the default `eventsQueue` subscribes to the given topic.
+   *
+   * @param extension `QueueExtension` added to the service
+   * @returns the queue subscribed to the given topic
+   */
+  public subscribe(extension: QueueExtension) : SubscriptionQueue {
+    let ret = {
+      queue: extension.eventsQueue,
+      scaleOnLatency: extension.autoscalingOptions,
+    } as SubscriptionQueue;
+
+    if (this.subscriptionQueue) {
+      ret = this.subscriptionQueue;
+    }
+    this.topic.addSubscription(new subscription.SqsSubscription(ret.queue));
+    return ret;
+  }
+}
+
+/**
+ * Settings for the hook which mutates the application container
+ * to add the events queue URI to its environment.
+ */
+interface ContainerMutatingProps {
+  /**
+   * The events queue name and URI to be added to the container environment.
+   */
+  readonly environment: { [key: string]: string };
+}
+
+/**
+ * This hook modifies the application container's environment to
+ * add the queue URL for the events queue of the service.
+ */
+class QueueExtensionMutatingHook extends ContainerMutatingHook {
+  private environment: { [key: string]: string };
+
+  constructor(props: ContainerMutatingProps) {
+    super();
+    this.environment = props.environment;
+  }
+
+  public mutateContainerDefinition(props: ecs.ContainerDefinitionOptions): ecs.ContainerDefinitionOptions {
+    return {
+      ...props,
+
+      environment: { ...(props.environment || {}), ...this.environment },
+    } as ecs.ContainerDefinitionOptions;
+  }
+}
+
+/**
+ * This extension creates a default `eventsQueue` for the service (if not provided) and accepts a list of objects of
+ * type `ISubscribable` that the `eventsQueue` subscribes to. It creates the subscriptions and sets up permissions
+ * for the service to consume messages from the SQS Queues.
+ *
+ * It also configures auto scaling for the service by scaling up or down to maintain an acceptable queue latency by
```

Comment:
I'm confused by "queue latency" in this context. Is this really the most precise term here? Shouldn't it be "queue size"? I am not at all an expert in this area, so will defer to you @upparekh.  (The same sentence is in the README as well, so if you make a change here, make sure to make it there.)

Reply:
I think 'queue latency' should be good. The acceptable backlog per task metric used as target value by the scaling policy is indicative of the maximum queue latency the service can tolerate. If the backlog per task increases beyond this value, more tasks are added.